crates/predictor/src/predictor/dataset_reader.rs

```diff
@@ -89,11 +90,30 @@ impl UninitializedDataSetReader {
                     }
                 };
 
-                let mut readers : Vec<Box<GribReader>> = vec![];
-                for file in bucket {
+                let mut readers : HashMap<i32, Box<GribReader>> = HashMap::new();
+                let mut last_hour = 0.0;
+                let mut last_file = &bucket[0];
+
+                for file in &bucket[1..] {
                     println!("{}", file.path().display());
 
-                    readers.push(Box::new(GribReader::new(file.path().to_str().unwrap().to_string())));
+                    let name = file.file_name().into_string().unwrap();
+                    let hour = name.split("_").collect::<Vec<&str>>()[4][0..3].parse::<f32>().unwrap();
+
+                    let divider = (hour - (((hour-last_hour)-1.0)/2.0)).round() as i32;
+
+                    for hr in (last_hour as i32)..divider {
+                        let reader = Box::new(GribReader::new(last_file.path().to_str().unwrap().to_string()));
```

Comment:
Same as below -- This creates a new grib reader for each hour, which will make it really inefficient down the road. Instead, you should just pass around a reference to the same reader.